gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java

```diff
@@ -145,7 +150,25 @@ public JsonWriter value(Number value) throws IOException {
         throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value);
       }
     }
-    generator.writeNumber(d);
+    if (value instanceof Integer) {
+      generator.writeNumber((Integer) value);
+    } else if (value instanceof Short) {
+      generator.writeNumber((Short) value);
+    } else if (value instanceof Long) {
+      generator.writeNumber((Long) value);
+    } else if (value instanceof Float) {
+      generator.writeNumber((Float) value);
+    } else if (value instanceof Double) {
+      generator.writeNumber((Double) value);
+    } else if (value instanceof BigInteger) {
+      generator.writeNumber((BigInteger) value);
+    } else if (value instanceof BigDecimal) {
+      generator.writeNumber((BigDecimal) value);
+    } else if (floor(d) == d) {
```

Comment:
Let me think more about cases when a double / float `1.0D` needs to be serialized as int `1`. 

One might potentially lose information like positive and negative zero `-0.0F` and `+0.0F`

Reply:
More than happy for the else branch to be double. Let me know what you want here

Reply:
Yes.  Right now I think double should be last resort (if no other types are matching) :
```java
// many instanceof checks 
} else {
  // last resort serialize as double
 generator.writeNumber(number.doubleValue());
}

```